RFC-0009-jdbc-join-push-down.md

```diff
@@ -0,0 +1,888 @@
+| 1  | abs(int_clumn) = int_cilumn2        | `Select * from table a join table b on abs(a.col1) = b.col2;`                            
+| 2  | int_sum_column = int_value1_column1 + int_value1_column2       | `Select * from table a join table b on a.col1 = b.col2 + b.col3;` 
+| 3  | cast(varchar_20_column, varchar(100)) = varchar100_column       | `Select * from table a join table b on cast(a.varchar_20_column, varchar(100)) = b.varchar100_column;` 
+
+
+**2. Join operation should be an INNER JOIN or a SELF JOIN.**
+
+A SELF JOIN is when a table is joined with itself.
+
+Note: Other optimizers in Presto may change the Join operation. We can call this as Inference. Sometimes presto will change a Pushdown capable Inner join to another Join operation incapable of pushdown (Eg: Infering to remove join condition/predicate in the plan). This will lead to pushdown capability being removed. And sometimes presto will change Join operation to a pushdown capable one. (Eg: Infering to create Inner join from Right/Left join)
+
+Examples to explain presto change an inner join to another Join operation : 
+
+Suppose we have a query like this: 
+
+`Select * from table a join table b on a.col1 = b.col2 and a.col1 = 5;`
+
+Presto will change this from an inner join to two different select statements like this: 
+
+`Select * from table a where a.col1 = 5;`
+
+`Select * from table b where b.col2 = 5;`
+
+Then it does a cross join with these two results. We will not do pushdown in this case.
+
+**3. Join criteria (joining column) should be of Datatypes and operators that support join pushdown.**
+
+| No | DataType support join pushdown                   | Operations                                           |
+|----|-------------------------------------|-------------------------------------------------------------------|
+| 1  | TinyINT        | `=, <, >, <=, >=, !=, <>`                    |
+| 2  | SmallINT       | `=, <, >, <=, >=, !=, <>`                    |   
+| 3  | Integer        | `=, <, >, <=, >=, !=, <>`                    |
+| 4  | BigINT         | `=, <, >, <=, >=, !=, <>`                    |
+| 5  | Boolean        | `=, !=, <>`                                  |   
+| 6  | Integer        | `=, <, >, <=, >=, !=, <>`                    |
+| 7  | Real           | `=, <, >, <=, >=, !=, <>`                    |   
+| 8  | Double         | `=, <, >, <=, >=, !=, <>`                    |
+| 9  | Decimal        | `=, <, >, <=, >=, !=, <>`                    |
+| 10 | Varchar        | `=, <, >, <=, >=, !=, <>`                    |
+| 11 | Char           | `=, <, >, <=, >=, !=, <>`                    |
+
+
+**4. All tables from same connector will be grouped based on above specifications and pushed down to underlying datasource.**
+
+**5. Enable presto Join pushdown capabilities by setting the session flag optimizer_inner_join_pushdown_enabled = true.**
+
+## Low level Design
```

Comment:
Please rework this section or remove it. A lot of this is implementation detail. What we should be highlighting is 
1. How we would group tables by connector using the same mechanism in ReorderJoins. Describe how we would limit the grouping to only join sources that are either `TableScan` or `Filter<-TableScan` or `Project<-Filter<-TableScan` 
2. Describe the updates to the SPI to add a table handle or other data structure to represent a 'group' of individual table handles that are joined
3. Describe how the connectors, and how the JDBC connector in particular can indicate that it wants to 'participate' in this creation of grouped-tables
4. Describe how the JDBC connector will un-wrap these grouped-tables and build the SQL necessary to achieve the JOIN condition that it would use when querying the JDBC source
5. Describe what the corner cases are - why we would need a way to add a table alias

Please create smaller sections for these as you see fit

Reply:
Additionally, please add a section dedicated to SPI changes - you can add classes, interfaces and types that are being added or removed, and add code-comments on them to describe how they would be used

Reply:
Removed some things in that section.. Working on the rest

Reply:
Added the changes in SPI module.

Reply:
1. To be added
2. This is added in the SPI module section
3. To be added
4. Described in 'Create Join Query in QueryBuilder' section
5. Made a new section and put self join under it.